File: frontend/src/api/events.ts
import { api } from './client'

export interface EventResponse {
  id: string
  title: string
  description: string | null
  sport: string
  location: string | null
  city: string
  date_start: string
  date_end: string
  organizer_id: string
  max_participants: number | null
  status: string
  participant_count: number
  created_at: string
  updated_at: string
}

export interface EventCreatePayload {
  title: string
  description?: string | null
  sport: string
  location?: string | null
  city: string
  date_start: string
  date_end: string
  max_participants?: number | null
}

export interface EventUpdatePayload {
  title?: string
  description?: string | null
  sport?: string
  location?: string | null
  city?: string
  date_start?: string
  date_end?: string
  max_participants?: number | null
  status?: string
}

export const eventsApi = {
  list: (sport?: string, city?: string, skip?: number, limit?: number) =>
    api.get<EventResponse[]>('/events', { params: { sport, city, skip, limit } }),

  create: (payload: EventCreatePayload) =>
    api.post<EventResponse>('/events', payload),

  get: (eventId: string) =>
    api.get<EventResponse>(`/events/${eventId}`),

  update: (eventId: string, payload: EventUpdatePayload) =>
    api.put<EventResponse>(`/events/${eventId}`, payload),

  delete: (eventId: string) =>
    api.delete<void>(`/events/${eventId}`),

  join: (eventId: string) =>
    api.post<EventResponse>(`/events/${eventId}/join`, {}),

  leave: (eventId: string) =>
    api.delete<void>(`/events/${eventId}/leave`),
}

File: backend/app/schemas/events.py
"""Pydantic schemas for events."""

from datetime import datetime

from pydantic import BaseModel, Field


class EventParticipantResponse(BaseModel):
    user_id: str
    status: str
    joined_at: datetime

    model_config = {"from_attributes": True}


class EventCreateRequest(BaseModel):
    title: str = Field(max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    sport: str = Field(max_length=50)
    location: str | None = Field(default=None, max_length=255)
    city: str = Field(max_length=100)
    date_start: datetime
    date_end: datetime
    max_participants: int | None = Field(default=None, ge=1)


class EventUpdateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    sport: str | None = Field(default=None, max_length=50)
    location: str | None = Field(default=None, max_length=255)
    city: str | None = Field(default=None, max_length=100)
    date_start: datetime | None = None
    date_end: datetime | None = None
    max_participants: int | None = Field(default=None, ge=1)
    status: str | None = Field(default=None, max_length=20)


class EventResponse(BaseModel):
    id: str
    title: str
    description: str | None
    sport: str
    location: str | None
    city: str
    date_start: datetime
    date_end: datetime
    organizer_id: str
    max_participants: int | None
    status: str
    participant_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EventDetailResponse(EventResponse):
    participants: list[EventParticipantResponse] = []
